core/src/main/java/org/apache/iceberg/avro/ValueWriters.java

```diff
@@ -492,4 +561,153 @@ protected Object get(IndexedRecord struct, int pos) {
+    }
+  }
+
+  public abstract static class MetricsAwareStringWriter<T extends Comparable<T>> extends ComparableWriter<T> {
+    public MetricsAwareStringWriter(int id) {
+      super(id);
+    }
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      // convert min/max to string to allow upper/lower bound truncation when gathering metrics,
+      // as in different implementations there's no guarantee that input to string writer will be char sequence
+      return metrics(Object::toString);
+    }
+  }
+
+  private abstract static class MetricsAwareByteArrayWriter extends MetricsAwareWriter<byte[]> {
+    MetricsAwareByteArrayWriter(int id) {
+      super(id, Comparators.unsignedByteArrays());
+    }
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      // convert min/max to byte buffer to allow upper/lower bound truncation when gathering metrics.
+      return metrics(ByteBuffer::wrap);
+    }
+  }
+
+  public abstract static class ComparableWriter<T extends Comparable<T>> extends MetricsAwareWriter<T> {
+    public ComparableWriter(int id) {
+      super(id, Comparable::compareTo);
+    }
+  }
+
+  /**
+   * A value writer wrapper that keeps track of column statistics (metrics) during writing.
+   *
+   * @param <T> Input type
+   */
+  public abstract static class MetricsAwareWriter<T> extends MetricsAwareTransformWriter<T, T> {
+    public MetricsAwareWriter(int id, Comparator<T> comparator) {
+      super(id, comparator, Function.identity());
+    }
+
+    /**
+     * Helper class to transform the input type when collecting metrics.
+     * The transform function converts the stats information from the specific type that the underlying writer
+     * understands to a more general type that could be transformed to binary following iceberg single-value
+     * serialization spec.
+     *
+     * @param func tranformation function
+     * @return a stream of field metrics with bounds converted by the given transformation
+     */
+    protected Stream<FieldMetrics> metrics(Function<T, ?> func) {
+      return Stream.of(new FieldMetrics<>(id, valueCount, nullValueCount, 0,
+          updateBound(min, func), updateBound(max, func)));
+    }
+
+    private <T3, T4> T4 updateBound(T3 bound, Function<T3, T4> func) {
+      return bound == null ? null : func.apply(bound);
+    }
+  }
+
+  /**
+   * A value writer wrapper that keeps track of column statistics (metrics) during writing, and accepts a
+   * transformation in its constructor.
+   * The transformation will apply to the input data to produce the type that the underlying writer accepts.
+   * Stats will also be tracked with the type after transformation.
+   *
+   * @param <T1> Input type
+   * @param <T2> Type after transformation
+   */
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  public abstract static class MetricsAwareTransformWriter<T1, T2> implements ValueWriter<T1> {
```

Comment:
We typically use `S` and `T` for input and output types in other transform classes.